design/ironic-debuggability-improvement.md

```diff
@@ -0,0 +1,108 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# ironic-debuggability-improvement
+
+## Status
+
+provisional
+
+## Summary
+
+Ironic and Ironic-inspector logs are currently hidden inside their respective
+containers and never shown to the end user by default.
+This document describes proposed changes in Ironic and Ironic-inspector
+logs handling to make debugging easier for the end user.
+
+## Motivation
+
+We need to make Ironic logs accessible to the end user.
+
+### Goals
+
+1. Make it easier to understand what happened when a deployment fails.
+2. Avoid old logs pile up and take all available space.
+
+### Non-Goals
+
+1. Modify Ironic logging events or log level.
+
+## Proposal
+
+The ironic-inspector-image includes a new script, ironic-inspection-log-watch,
+that can act as an entry point for a container to dump host inspection logs.
+
+The script should watch for Ironic host inspection log files to appear in
+"/shared/logs/ironic-inspector/ramdisk", wait for them to stop changing
+(to ensure the entire log is dumped), decompress them, print their contents
```

Comment:
Could we note that the logs aren't streamed but rather uploaded once? 

We mentioned elsewhere that it was difficult to know when it would stop changing in the streaming case, but it's not a concern if it all happens in one shot.

Reply:
Yes, I think that would be a good addition, especially for anyone (like myself) who isn't intimately familiar with Ironic's implementation details. We should describe when we expect log files to appear (after inspection and after provisioning are implied, but maybe also after deprovisioning?), and which service is going to be writing the files.

I assume there is some way to look at the filename and determine which node the log is for? Are they named with the UUID, for example?

Reply:
@dtantsur , @zaneb - do you have any insights on these logs behaviour?

Reply:
File names for ironic-inspector are configurable, this is the default: https://opendev.org/openstack/ironic-inspector/src/branch/master/ironic_inspector/conf/processing.py#L87-L94. The logs are written all at once, which is not necessary atomic, but pretty close to. The ironic file names are a bit dynamic (https://opendev.org/openstack/ironic/src/branch/master/ironic/drivers/utils.py#L266-L282) but they'll also start with a UUID and end with a timestamp.

Reply:
So with the current configuration we can find the files in the log by ironic uuid, e.g:

```
# 7a72d29d-f99a-4322-b41a-d8d1340d1009 is a node UUID retrieved from ironic
$ find / -name "7a72d29d-f99a-4322-b41a-d8d1340d1009*"
/shared/log/ironic/deploy/7a72d29d-f99a-4322-b41a-d8d1340d1009_01ed04b3-22c5-4b57-b7b6-b2e31b7089fc_2020-10-20-14-32-46.tar.gz
/shared/log/ironic-inspector/ramdisk/7a72d29d-f99a-4322-b41a-d8d1340d1009_20201020-143110.478287.tar.gz
```

It's a shame that ironic doesn't offer the option to include the name in the filename format, as that would make the log output a bit more easily searchable (particularly in the case where the metal3 pod gets rescheduled so all the UUIDs change?)

Reply:
Yeah, we need *something* to tie the logs back to the host in kubernetes. The ironic UUID is 2 hops (and ephemeral as you point out), but the name should match the host resource name. Can the script that cats out the log files use the UUID in the log filename to look up the node's name, then include that value as a prefix for each line of the output?

Reply:
Looking at ironic documentation, something like `ironic node-show --fields name <node UUID>` should do the trick. But I need to verify.

Reply:
It works from the provisioning host:
```
[akiselev@nfvsdn-07 metal3-dev-env]$ baremetal node show d712c4a4-eaa4-4767-8c45-feb7f392ddda --fields name
Management cluster forced to be minikube when container runtime is not docker
+-------+--------+
| Field | Value  |
+-------+--------+
| name  | node-1 |
+-------+--------+
```

Reply:
Are the client tools in the ironic image? And is there a clouds.yaml set up?

Reply:
I don't see Ironic client tools in either image. Neither kubectl (another alternative) is there. @elfosardo , @derekhiggins  - what would it take to add these tools into ironic and ironic-inspector images?

Reply:
Adding the ironic client shouldn't be difficult but giving it access to the credentials would be more involved, I guess they would need to be added to a volume and mounted into the container somehow.

Reply:
Thank you. @elfosardo suggested to use https://github.com/metal3-io/ironic-client instead, but credentials storage question is still open.

Reply:
We give the baremetal-operator access to the ironic credentials. Would we need to do anything differently for this tool?

Reply:
the same credentials could be passed in, we'd just have to write them to a clouds.yaml file

Reply:
> "It's a shame that ironic doesn't offer the option to include the name in the filename format, as that would make the log output a bit more easily searchable (particularly in the case where the metal3 pod gets rescheduled so all the UUIDs change?)"

Just noting that the "Instance UUID" field is part of the filename for the deploy logs (not inspector), which is the bmh uid.

Reply:
Can you elaborate more how to pass ironic credentials to ironic-client, maybe a link to BMO code?

Reply:
> > "It's a shame that ironic doesn't offer the option to include the name in the filename format, as that would make the log output a bit more easily searchable (particularly in the case where the metal3 pod gets rescheduled so all the UUIDs change?)"
> 
> Just noting that the "Instance UUID" field is part of the filename for the deploy logs (not inspector), which is the bmh uid.

So you're saying the filenames for inspection logs do not include the UUID? That makes sense, since the inspector may not know the UUID (especially if the node does not exist in ironic when inspection runs).

Reply:
To use the ironic client image metal3-dev-env has a wrapper script, [here](https://github.com/metal3-io/metal3-dev-env/blob/master/openstackclient.sh) 
It expects the clouds.yaml file that Doug mentioned to be in the mount directory _clouds_yaml/clouds.yaml, dev-scripts creates this if you want to see an example
Finally, if for some reason the clouds.yaml file wont work you can use env variables
[derekh@r640-u07 dev-scripts]$ export OS_PASSWORD=XXXX
[derekh@r640-u07 dev-scripts]$ podman run --net=host --rm -e OS_ENDPOINT=http://[fd00:1101::3]:6385 -e OS_USERNAME=ironic-user -e OS_PASSWORD --entrypoint baremetal quay.io/metal3-io/ironic-client --os-auth-type http_basic node list

Reply:
@derekhiggins Can you point me to dev-scripts creating clouds.yaml code? Should I try custom metal3-dev-env image with ironic-client first, or we are confident in the above design for now?

Reply:
The code that creates the file is here
https://github.com/openshift-metal3/dev-scripts/blob/079a6ce/utils.sh#L269
here is an example one
https://goodsquishy.com/upload/725c0f07a367a04852af

If we end up using the ironic-client image then we'd just need to add your log-watch script and use it as an entry point.

Reply:
I think the reason we were going with the ironic-image is that it's already included in our release payload so adding another container to the pod is trivial. Is the client image included in that payload already, too?

Reply:
I talked to @juliakreger yesterday and opened two new stories on storyboard.openstack.org to add node name into log file names. I would be still curious to know if ironic-client image is included  into the payload once @derekhiggins is back.

Reply:
> I think the reason we were going with the ironic-image is that it's already included in our release payload so adding another container to the pod is trivial. Is the client image included in that payload already, too?

Agreed, Including ironic client in the ironic-image  would certainly make things simpler